From a333eaa3f0fbce0c77cea23015144e13d9c039e2 Mon Sep 17 00:00:00 2001
Message-ID: <a333eaa3f0fbce0c77cea23015144e13d9c039e2.1747030627.git.pmatilai@redhat.com>
From: Panu Matilainen <pmatilai@redhat.com>
Date: Wed, 23 Apr 2025 13:03:30 +0300
Subject: [PATCH] Fix a refactoring regression in rich dep and rpmdb set
 calculations

Commit 52f311e224254d7cb28bf2ced22db7b50e1a2653 failed to take into
account that std::set_difference() and std::set_intersection() require
both data sets to be sorted, whereas the old manual implementation only
required the second set to be sorted.

Add a minimal testcase for the WITH rich dep case from the original
report. Technically the same issue could occur with WITHOUT rich
dependencies and rpmdb lookups through dbiFindMatches() but I don't
reproducers for those so just fixing the code for the "obvious" error.

I'm not 100% confident this addresses all such assumptions in the
codebase. dbiIndexSet should really use a data structure that doesn't
require manual sorting like this, but then that's a much bigger change
as the rpmdb code is currently accessing the set members via an index...

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2360342
---
 lib/backend/dbiset.cc |  4 ++++
 lib/backend/dbiset.hh |  6 ++++--
 tests/rpmdeps.at      | 10 ++++++++--
 3 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/lib/backend/dbiset.cc b/lib/backend/dbiset.cc
index 14b6d7857..e29dfd377 100644
--- a/lib/backend/dbiset.cc
+++ b/lib/backend/dbiset.cc
@@ -101,6 +101,8 @@ int dbiIndexSetPruneSet(dbiIndexSet set, dbiIndexSet oset, int sorted)
     if (set->recs.empty() || oset == NULL || oset->recs.empty())
 	return 1;
 
+    if (set->recs.size() > 1 && !sorted)
+	dbiIndexSetSort(set);
     if (oset->recs.size() > 1 && !sorted)
 	dbiIndexSetSort(oset);
 
@@ -120,6 +122,8 @@ int dbiIndexSetFilterSet(dbiIndexSet set, dbiIndexSet oset, int sorted)
 	return num ? 0 : 1;
     }
 
+    if (set->recs.size() > 1 && !sorted)
+	dbiIndexSetSort(set);
     if (oset->recs.size() > 1 && !sorted)
 	dbiIndexSetSort(oset);
 
diff --git a/lib/backend/dbiset.hh b/lib/backend/dbiset.hh
index fdb28e96d..10d584545 100644
--- a/lib/backend/dbiset.hh
+++ b/lib/backend/dbiset.hh
@@ -51,9 +51,10 @@ int dbiIndexSetAppendOne(dbiIndexSet set, unsigned int hdrNum,
 
 /**
  * Remove an index set from another.
+ * The sets are sorted as a side-effect if sorted is 0.
  * @param set          set of index database items
  * @param oset         set of entries that should be removed
- * @param sorted       oset is already sorted?
+ * @param sorted       are set and oset already sorted?
  * @return             0 success, 1 failure (no items found)
  */
 RPM_GNUC_INTERNAL
@@ -61,9 +62,10 @@ int dbiIndexSetPruneSet(dbiIndexSet set, dbiIndexSet oset, int sorted);
 
 /**
  * Filter (intersect) an index set with another.
+ * The sets are sorted as a side-effect if sorted is 0.
  * @param set          set of index database items
  * @param oset         set of entries that should be intersected
- * @param sorted       oset is already sorted?
+ * @param sorted       are set and oset already sorted?
  * @return             0 success, 1 failure (no items removed)
  */
 RPM_GNUC_INTERNAL
diff --git a/tests/rpmdeps.at b/tests/rpmdeps.at
index 4aca7168e..469ea04db 100644
--- a/tests/rpmdeps.at
+++ b/tests/rpmdeps.at
@@ -327,17 +327,23 @@ AT_KEYWORDS([install, boolean])
 runroot rpmbuild --quiet -bb \
 	--define "pkg one" \
 	--define "reqs (deptest-two with flavor = desktop)" \
+	--define "provs feat = 4.0" \
 	  /data/SPECS/deptest.spec
 
 runroot rpmbuild --quiet -bb \
 	--define "pkg two" \
-	--define "provs flavor = desktop" \
+	--define "provs flavor = desktop feat = 1.5" \
+	  /data/SPECS/deptest.spec
+
+runroot rpmbuild --quiet -bb \
+	--define "pkg three" \
+	--define "reqs (feat >= 1.0 with feat < 3.0)" \
 	  /data/SPECS/deptest.spec
 
 # satisfied WITH require in transaction
 RPMTEST_CHECK([
 RPMDB_RESET
-runroot rpm -U /build/RPMS/noarch/deptest-one-1.0-1.noarch.rpm /build/RPMS/noarch/deptest-two-1.0-1.noarch.rpm
+runroot rpm -U /build/RPMS/noarch/deptest-two-1.0-1.noarch.rpm /build/RPMS/noarch/deptest-three-1.0-1.noarch.rpm /build/RPMS/noarch/deptest-one-1.0-1.noarch.rpm
 ],
 [0],
 [],
-- 
2.49.0

